File: 백준/Gold/1707. 이분 그래프/이분 그래프.js
class Queue {
  constructor() {
    this.items = {};
    this.headIndex = 0;
    this.tailIndex = 0;
  }
  enqueue(item) {
    this.items[this.tailIndex] = item;
    this.tailIndex++;
  }
  dequeue() {
    const item = this.items[this.headIndex];
    delete this.items[this.headIndex];
    this.headIndex++;
    return item;
  }
  peek() {
    return this.items[this.headIndex];
  }
  getLength() {
    return this.tailIndex - this.headIndex;
  }
}

function toggle_flag(flag) {
  if (flag == 1) {
    return 2;
  } else {
    return 1;
  }
}

function dfs(x) {

  for (next of graph[x]) {
    if (!visited[next]) {
      visited[next] = toggle_flag(visited[x]);
      dfs(next);
    }
  }
}

function check_bipartite_graph(v) {
  for (let i = 1; i < v + 1; i++) {
    if (graph[i]) {
      for (node of graph[i]) {
        if (visited[i] == visited[node]) {
          return false;
        }
      }
    }
  }
  return true;
}

let fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().split("\n");

let k = Number(input[0]);
let graph;
let visited;
let line = 1;
while (k) {
  const [v, e] = input[line++].split(" ").map(Number);
  graph = Array.from({ length: v + 1 }, () => []);
  visited = Array.from({ length: v + 1 }, () => false);
  for (let i = 0; i < e; i++) {
    const [u, v] = input[line++].split(" ").map(Number);
    graph[u].push(v);
    graph[v].push(u);
  }
  for (let i = 1; i < v + 1; i++) {
    if (!visited[i]) {
      visited[i] = 1;
      dfs(i);
    }
  }
  if (check_bipartite_graph(v)) {
    console.log("YES");
  } else {
    console.log("NO");
  }
  k--;
}

File: 백준/Gold/1707. 이분 그래프/이분 그래프.py
import sys
from collections import deque

input = sys.stdin.readline

def bfs(i):
    q= deque()
    q.append(i)
    visited[i] = 0

    while q:
        cur = q.popleft()

        for next in graph[cur]:
            if visited[next] == -1:
                visited[next] = (visited[cur]+1)%2
                q.append(next)
            elif visited[next] == visited[cur]:  # 같은 색이라면 이분 그래프가 아님
                return False
    return True

K = int(input())
for _ in range(K):
    V, E = map(int, input().split())
    graph = [[] for _ in range(V+1)]
    visited = [-1 for _ in range(V+1)]
    for _ in range(E):
        u,v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)

    answer = True
    for i in range(1, V+1):
        if visited[i] == -1:
            if not bfs(i):
                answer = False
                break

    if answer == True:
        print('YES')
    else:
        print('NO')
